src/nvim/lua/executor.c

```diff
@@ -0,0 +1,576 @@
+  char *lcmd;
+  if (lcmd_len < IOSIZE) {
+    lcmd = (char *)IObuff;
+  } else {
+    lcmd = xmalloc(lcmd_len);
+  }
+  memcpy(lcmd, EVALHEADER, sizeof(EVALHEADER) - 1);
+  memcpy(lcmd + sizeof(EVALHEADER) - 1, str->data, str->size);
+  lcmd[lcmd_len - 1] = ')';
+#undef EVALHEADER
+  if (luaL_loadbuffer(lstate, lcmd, lcmd_len, NLUA_EVAL_NAME)) {
+    nlua_error(lstate,
+               _("E5107: Error while creating lua chunk for luaeval(): %.*s"));
+    if (lcmd != (char *)IObuff) {
+      xfree(lcmd);
+    }
+    return 0;
+  }
+  if (lcmd != (char *)IObuff) {
+    xfree(lcmd);
+  }
+
+  if (arg == NULL || arg->v_type == VAR_UNKNOWN) {
+    lua_pushnil(lstate);
+  } else {
+    nlua_push_typval(lstate, arg);
+  }
+  if (lua_pcall(lstate, 1, 1, 0)) {
+    nlua_error(lstate,
+               _("E5108: Error while calling lua chunk for luaeval(): %.*s"));
+    return 0;
+  }
+  if (!nlua_pop_typval(lstate, ret_tv)) {
+    return 0;
+  }
+
+  return 0;
+}
+
+/// Print as a Vim message
+///
+/// @param  lstate  Lua interpreter state.
+static int nlua_print(lua_State *const lstate)
+  FUNC_ATTR_NONNULL_ALL
+{
+#define PRINT_ERROR(msg) \
+  do { \
+    errmsg = msg; \
+    errmsg_len = sizeof(msg) - 1; \
+    goto nlua_print_error; \
+  } while (0)
+  const int nargs = lua_gettop(lstate);
+  lua_getglobal(lstate, "tostring");
+  const char *errmsg = NULL;
+  size_t errmsg_len = 0;
+  garray_T msg_ga;
+  ga_init(&msg_ga, 1, 80);
+  int curargidx = 1;
+  for (; curargidx <= nargs; curargidx++) {
+    lua_pushvalue(lstate, -1);  // tostring
+    lua_pushvalue(lstate, curargidx);  // arg
+    if (lua_pcall(lstate, 1, 1, 0)) {
+      errmsg = lua_tolstring(lstate, -1, &errmsg_len);
+      goto nlua_print_error;
+    }
+    size_t len;
+    const char *const s = lua_tolstring(lstate, -1, &len);
+    if (s == NULL) {
+      PRINT_ERROR(
+          "<Unknown error: lua_tolstring returned NULL for tostring result>");
+    }
+    ga_concat_len(&msg_ga, s, len);
+    if (curargidx < nargs) {
+      ga_append(&msg_ga, ' ');
+    }
+    lua_pop(lstate, 1);
+  }
+#undef PRINT_ERROR
+  lua_pop(lstate, nargs + 1);
+  ga_append(&msg_ga, NUL);
+  {
+    const size_t len = (size_t)msg_ga.ga_len - 1;
+    char *const str = (char *)msg_ga.ga_data;
+
+    for (size_t i = 0; i < len;) {
+      const size_t start = i;
+      while (i < len) {
+        switch (str[i]) {
+          case NUL: {
+            str[i] = NL;
+            i++;
+            continue;
+          }
+          case NL: {
+            str[i] = NUL;
+            i++;
+            break;
+          }
+          default: {
+            i++;
+            continue;
+          }
+        }
+        break;
+      }
+      msg((char_u *)str + start);
+    }
+    if (str[len - 1] == NUL) {  // Last was newline
+      msg((char_u *)"");
+    }
+  }
+  ga_clear(&msg_ga);
+  return 0;
+nlua_print_error:
+  emsgf(_("E5114: Error while converting print argument #%i: %.*s"),
+        curargidx, errmsg_len, errmsg);
```

Comment:
Length argument should be of type int: `errmsg_len` -> `(int)errmsg_len`, otherwise macros `va_*()` parse the parameter list wrong.